src/pcre2_internal.h

```diff
@@ -574,6 +574,13 @@ modes. */
 #define REQ_CU_MAX       2000
 #endif
 
+/* The maximum nesting depth for Unicode character class sets.
+Currently fixed. Warning: the interpreter relies on this so it can encode
+the operand stack in a uint32_t. A nesting limit of 15 implies (15*2+1)=31
+stack operands required. */
+
+#define ECLASS_NEST_LIMIT  15
```

Comment:
I thought 1 bit is enough for an xclass result value. Why do you need two?

Reply:
Aha, because you can have two operators within a '[...]'. So if you have two brackets: `[A--B[C--DE]]` then the worst-case and peak stack usage is that the leafmost `[C--DE]` takes three slots on the stack (result of C, D and E, before applying || to DE and --), and the non-leafmost `[A--B ...foo...]` takes two slots on the stack (need to store result of matching A and B).

Hence 5 stack slots for two `[ ... [...] ]` brackets. It's 2n+1.

It would be worse if we allowed mixing precedence within the brackets; then each `[...]` could consume even more stack slots.

Reply:
I thought the pushes/pops are tracked individually. Something like `/[A]&&[B]||[C]--[D]/` is `push [A], push [B], &&, push [C], ||, push[D], --` and the max stack size is 2.

Recursive descent should be able to compute it. This looks like another thing to do then.

Reply:
For your example (`/[A]&&[B]||[C]--[D]/`), yes, the maximum stack size is 2.

The worst case is a stack size of three, if you have `A op1 B op2 C` where op2 is higher (tighter) precedence.

We have a hard limit of 15 nested brackets; which gives us a maximum stack consumption of 31 slots.

Of course we can calculate (if we want) what the actual stack slots consumed will be, but we guarantee it will fit into a uint32_t if we limit the brackets to at most 15 deep.